File: packages/imanikurd/src/companions.ts
import { loadData, loadDataAsync } from "./loadData.js";

export interface Companion {
  id: number;
  name: string;
  arabic?: string;
  description?: string;
  [key: string]: unknown;
}

let companionsCache: Companion[] | null = null;

function getCompanionsData(): Companion[] {
  if (!companionsCache) {
    companionsCache = loadData<Companion[]>("companions.json");
  }
  return companionsCache;
}

async function getCompanionsDataAsync(): Promise<Companion[]> {
  if (!companionsCache) {
    companionsCache = await loadDataAsync<Companion[]>("companions.json");
  }
  return companionsCache;
}

/** Get all companions (Sahabah) biographies */
export function getCompanions(): Companion[] {
  return getCompanionsData();
}

/** Get all companions (Sahabah) biographies asynchronously */
export async function getCompanionsAsync(): Promise<Companion[]> {
  return getCompanionsDataAsync();
}

/** Get a companion by id */
export function getCompanion(id: number): Companion | undefined {
  return getCompanionsData().find((c) => c.id === id);
}

/** Get a companion by id asynchronously */
export async function getCompanionAsync(id: number): Promise<Companion | undefined> {
  const data = await getCompanionsDataAsync();
  return data.find((c) => c.id === id);
}

/** Search companions by name */
export function searchCompanions(query: string): Companion[] {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  return getCompanionsData().filter((c) => {
    const name = String(c.name || "").toLowerCase();
    const arabic = String(c.arabic || "");
    const desc = String(c.description || "").toLowerCase();
    return name.includes(normalized) || arabic.includes(query.trim()) || desc.includes(normalized);
  });
}

/** Search companions by name asynchronously */
export async function searchCompanionsAsync(query: string): Promise<Companion[]> {
  const normalized = query.trim().toLowerCase();
  if (!normalized) return [];

  const data = await getCompanionsDataAsync();
  return data.filter((c) => {
    const name = String(c.name || "").toLowerCase();
    const arabic = String(c.arabic || "");
    const desc = String(c.description || "").toLowerCase();
    return name.includes(normalized) || arabic.includes(query.trim()) || desc.includes(normalized);
  });
}

File: packages/imanikurd_flutter/lib/src/companions.dart
import 'load_data.dart';
import 'models.dart';

List<Companion>? _companionsCache;

Future<List<Companion>> _getCompanionsData() async {
  _companionsCache ??= (await loadDataAsync<List<dynamic>>('companions.json'))
      .map((e) => Companion.fromJson(e as Map<String, dynamic>))
      .toList();
  return _companionsCache!;
}

Future<List<Companion>> getCompanions() => _getCompanionsData();

Future<Companion?> getCompanion(int id) async {
  final data = await _getCompanionsData();
  for (final companion in data) {
    if (companion.id == id) return companion;
  }
  return null;
}

Future<List<Companion>> searchCompanions(String query) async {
  final normalized = query.trim().toLowerCase();
  if (normalized.isEmpty) return [];
  final data = await _getCompanionsData();
  return data.where((c) {
    final name = c.name.toLowerCase();
    final arabic = c.arabic ?? '';
    final desc = (c.description ?? '').toLowerCase();
    return name.contains(normalized) ||
        arabic.contains(query.trim()) ||
        desc.contains(normalized);
  }).toList();
}
